File: hahahaeun/프로그래머스/0/120866. 안전지대/안전지대.py
def solution(board):
    answer = 0
    spot = []
    n = len(board)
    check = [[True for _ in range(n)] for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            if board[i][j] == 1:
                spot.append((j,i))
                
    around = [(0,1),(0,-1),(1,0),(-1,0),(1,-1),(-1,1),(-1,-1),(1,1)]
    
    for i in spot:
        check[i[1]][i[0]] = False
        for j in range(8):
            nx = i[0] + around[j][0]
            ny = i[1] + around[j][1]

            if 0<=nx<n and 0<=ny<n and check[ny][nx]:
                check[ny][nx] = False
    
    for i in range(n):
        for j in range(n):
            if check[i][j]:
                answer += 1
            
    return answer

File: limyj108/프로그래머스/0/120866. 안전지대/안전지대.cpp
#include <string>
#include <vector>

using namespace std;

pair<int, int> drc[8] = {{-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}};

bool is_valid(pair<int, int> p, int board_rows, int board_cols);

int solution(vector<vector<int>> board) {
    int answer = 0;
    int board_rows = board.size(), board_cols = board[0].size();
    
    for (int r = 0; r < board_rows; r++) {
        for (int c = 0; c < board_cols; c++) {
            bool is_safe = board[r][c] == 0;
            for (pair<int, int> d_ : drc) {
                pair<int, int> next = {r + d_.first, c + d_.second};
                if (is_valid(next, board_rows, board_cols) && board[next.first][next.second] == 1) {
                    is_safe = false;
                }
            }
            if (is_safe) {
                answer++;
            }
        }
    }
    return answer;
}

bool is_valid(pair<int, int> p, int board_rows, int board_cols) {
    return 0 <= p.first && p.first < board_rows && 0 <= p.second && p.second < board_cols;
}